pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1151 @@
+        let _was_first = match self {
+            ResidentOrWantedEvicted::Resident(strong) => {
+                let weak = Arc::downgrade(strong);
+                *self = ResidentOrWantedEvicted::WantedEvicted(weak);
+                // returning the weak is not useful, because the drop could had already ran with
+                // the replacement above, and that will take care of cleaning the Option we are in
+                true
+            }
+            ResidentOrWantedEvicted::WantedEvicted(_) => false,
+        };
+
+        match self {
+            ResidentOrWantedEvicted::WantedEvicted(ref weak) => weak,
+            _ => unreachable!("just wrote wanted evicted"),
+        }
+    }
+}
+
+struct LayerInner {
+    /// Only needed to check ondemand_download_behavior_treat_error_as_warn and creation of
+    /// [`Self::path`].
+    conf: &'static PageServerConf,
+    path: PathBuf,
+
+    desc: PersistentLayerDesc,
+
+    /// Timeline access is needed for metrics.
+    timeline: Weak<Timeline>,
+
+    /// Cached knowledge of [`Timeline::remote_client`] being `Some`.
+    have_remote_client: bool,
+
+    access_stats: LayerAccessStats,
+
+    /// This custom OnceCell is backed by std mutex, but only held for short time periods.
+    /// Initialization and deinitialization is done while holding a permit.
+    inner: heavier_once_cell::OnceCell<ResidentOrWantedEvicted>,
+
+    /// Do we want to garbage collect this when `LayerInner` is dropped, where garbage collection
+    /// means:
+    /// - schedule remote deletion
+    /// - instant local deletion
+    wanted_garbage_collected: AtomicBool,
+
+    /// Do we want to evict this layer as soon as possible? After being set to `true`, all accesses
+    /// will try to downgrade [`ResidentOrWantedEvicted`], which will eventually trigger
+    /// [`LayerInner::on_drop`].
+    wanted_evicted: AtomicBool,
+
+    /// Version is to make sure we will in fact only evict a file if no new guard has been created
+    /// for it.
+    version: AtomicUsize,
+
+    /// Allow subscribing to when the layer actually gets evicted.
+    status: tokio::sync::broadcast::Sender<Status>,
+    consecutive_failures: AtomicUsize,
+}
+
+impl std::fmt::Display for LayerInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.layer_desc().short_id())
+    }
+}
+
+impl AsLayerDesc for LayerInner {
+    fn layer_desc(&self) -> &PersistentLayerDesc {
+        &self.desc
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Status {
+    Evicted,
+    Downloaded,
+}
+
+impl Drop for LayerInner {
+    fn drop(&mut self) {
+        if !*self.wanted_garbage_collected.get_mut() {
+            // should we try to evict if the last wish was for eviction?
+            // feels like there's some hazard of overcrowding near shutdown near by, but we don't
+            // run drops during shutdown (yet)
+            return;
+        }
```

Comment:
We have never run drops on shutdown because we do `std::process::exit(0)`. Drops only get run on tenants which get detached, deleted (unsure) or timelines which get removed. I've been trying to change this..

Similarly this is why we sometimes seemingly leak walredo processes for non-removed tenants. systemd kills them in staging and prod for us.

I'd say executing the last wish could only conflict with the current `std::fs::remove_dir_all`, so this wouldn't be my first step to fixing this. Of course, adding in "run drops on shutdown" this late to the game might not be unproblematic either...